core/src/commonMain/kotlin/com/ustadmobile/core/controller/ContentEntryEdit2Presenter.kt

```diff
@@ -0,0 +1,165 @@
+        override fun toString(): String {
+            return storage.label
+        }
+    }
+
+    class LicenceMessageIdOptions(licence: LicenceOptions,context: Any)
+        : MessageIdOption(licence.messageId,context, licence.optionVal)
+
+    override val persistenceMode: PersistenceMode
+        get() = PersistenceMode.DB
+
+    private var storageOptions: List<UMStorageDir>? = null
+
+    override fun onCreate(savedState: Map<String, String>?) {
+        super.onCreate(savedState)
+        view.licenceOptions = LicenceOptions.values().map { LicenceMessageIdOptions(it, context) }
+        parentEntryUid = arguments[ARG_PARENT_ENTRY_UID]?.toLong()?:0
+        systemImpl.getStorageDirs(context, object : UmResultCallback<List<UMStorageDir>> {
+            override fun onDone(result: List<UMStorageDir>?) {
+                storageOptions = result
+                if(result != null){
+                   view.runOnUiThread(Runnable {
+                       view.setUpStorageOptions(result)
+                   })
+                }
+            }
+        })
+    }
+
+    override suspend fun onLoadEntityFromDb(db: UmAppDatabase): ContentEntryWithLanguage? {
+        val entityUid = arguments[ARG_ENTITY_UID]?.toLong() ?: 0
+        return withTimeoutOrNull(2000) {
+            db.contentEntryDao.findEntryWithLanguageByEntryId(entityUid)
+        } ?: ContentEntryWithLanguage()
+    }
+
+    override fun onLoadFromJson(bundle: Map<String, String>): ContentEntryWithLanguage? {
+        super.onLoadFromJson(bundle)
+        val entityJsonStr = bundle[ARG_ENTITY_JSON]
+        var editEntity: ContentEntryWithLanguage? = null
+        editEntity = if(entityJsonStr != null) {
+            Json.parse(ContentEntryWithLanguage.serializer(), entityJsonStr)
+        }else {
+            ContentEntryWithLanguage()
+        }
+        return editEntity
+    }
+
+    override fun onSaveInstanceState(savedState: MutableMap<String, String>) {
+        super.onSaveInstanceState(savedState)
+        val entityVal = view.entity
+        savedState.putEntityAsJson(ARG_ENTITY_JSON, null, entityVal)
+    }
+
+
+    override fun handleClickSave(entity: ContentEntryWithLanguage) {
+        GlobalScope.launch(doorMainDispatcher()) {
+            val canCreate = entity.title != null && ((entity.language != null
+                    && entity.licenseType != 0 && entity.leaf) || (!entity.leaf && entity.description != null))
+
+            if(canCreate){
+                if(entity.leaf){
+                    entity.licenseName = systemImpl.getString(LicenceOptions.values()
+                            .single { it.optionVal == entity.licenseType }.messageId, context)
+                }
+
+                if(entity.contentEntryUid == 0L) {
+                    entity.contentEntryUid = repo.contentEntryDao.insertAsync(entity)
+                    val contentEntryJoin = ContentEntryParentChildJoin()
```

Comment:
In Kotlin this could be written as:

val contentEntryJoin = ContentEntryParentChildJoin().apply {
    cepcjChildContentEntryUid = entity.contentEntryUid
    cepcjParentContentEntryUid = parentEntryUid
}